### ISSUE TITLE
False positive: E0602 undefined variable with assignment expression

### ISSUE THREAD

**Issue by jamesohortle** [PyCQA/pylint#5360]
### Bug description

The following snippet produces `E0602: Undefined variable 'x_0' (undefined-variable)` as output from pylint.

```py
sorted_things = sorted(
        things,
        key=lambda thing: x_0 
        if (x_0 := thing.this_value)
        < (x_1 := thing.that_value)
        else x_1,
)
```

### Command used

```shell
Basic `pylint` call as a pre-commit hook.
```

### Pylint output

```shell
pylint...................................................................Failed
- hook id: pylint
- exit code: 2

************* Module utils
project_dir/utils.py:197:22: E0602: Undefined variable 'x_0' (undefined-variable)
```

### Expected behavior

Should be no error.

### Pylint version

```shell
pylint --version
pylint 2.11.1
astroid 2.8.4
Python 3.9.5 (default, May  4 2021, 03:36:27) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

MacOS 11.5.2
